torch/distributions/multinomial.py

```diff
@@ -0,0 +1,91 @@
+import torch
+from torch.distributions.distribution import Distribution
+from torch.autograd import Variable
+from torch.distributions import Categorical
+from numbers import Number
+from torch.distributions import constraints
+from torch.distributions.utils import log_sum_exp, broadcast_all
+
+
+class Multinomial(Distribution):
+    r"""
+    Creates a Multinomial distribution parameterized by `total_count` and
+    either `probs` or `logits` (but not both). The innermost dimension of
+    `probs` indexes over categories. All other dimensions index over batches.
+
+    Note that `total_count` need not be specified if only :meth:`log_prob` is
+    called (see example below)
+
+    -   :meth:`sample` requires a single shared `total_count` for all
+        parameters and samples.
+    -   :meth:`log_prob` allows different `total_count` for each parameter and
+        sample.
+
+    Example::
+
+        >>> m = Multinomial(10, torch.Tensor([ 1, 1, 1, 1]))
+        >>> x= m.sample()  # equal probability of 0, 1, 2, 3
+         1
+         3
+         3
+         3
+        [torch.FloatTensor of size 4]
+
+        >>> Multinomial(probs=torch.Tensor([1, 1, 1, 1])).log_prob(x)
+        -4.1338
+        [torch.FloatTensor of size 1]
+
+    Args:
+        total_count (int or Tensor or Variable): number of trials
+        probs (Tensor or Variable): event probabilities
+        logits (Tensor or Variable): event log probabilities
+    """
+    params = {'total_count': constraints.nonnegative_integer,
+              'probs': constraints.simplex, 'logits': constraints.real}
+
+    def __init__(self, total_count=1, probs=None, logits=None):
+        if isinstance(total_count, Number):
+            self.total_count = total_count
+        else:
+            if isinstance(total_count, Variable):
+                total_count = total_count.data
+            self.total_count = total_count.view(-1)[0]
+            if (total_count != self.total_count).any():
```

Comment:
wouldn't it be simpler to disallow passing `total_count` vectors at all (just check that `total_count.numel() == 1`)?

Reply:
Done.